player_generic.gd

```diff
@@ -0,0 +1,270 @@
+			$Sprite.flip_h = false
+			continue
+		MOVELEFT_STATE:
+			$Sprite.flip_h = true
+			continue
+		COLLIDE_STATE:
+			print("COLLIDE_STATE")
+
+
+
+
+
+
+func _input (event):
+	if (Input.is_action_pressed ("move_left")):
+		(States &= ~IDLE_STATE)
+		(States |= MOVELEFT_STATE)
+		print(States)
+	if (Input.is_action_pressed ("move_right")):
+		(States &= ~IDLE_STATE)
+		(States |= MOVERIGHT_STATE)
+		print(States)
+	if (Input.is_action_just_released("move_left")):
+		(States &= ~MOVELEFT_STATE)
+		(States |= IDLE_STATE)
+		print(States)
+	if (Input.is_action_just_released("move_right")):
+		(States &= ~MOVERIGHT_STATE)
+		(States |= IDLE_STATE)
+		print(States)
+	if (OS.is_debug_build ()):
+		if (Input.is_action_just_pressed ("DEBUG_kill_player")):
+			print ("AAAA")
+			game_space.lives -= 1
+	return
+
+func _physics_process (delta):
+	CheckStateToPlaySprite(States)
+	if (horizontal_lock_timer >= 0):
+		horizontal_lock_timer -= delta
+	#	printerr ($FloorDetect.is_colliding ())
+	#	printerr ($FloorDetect.get_collision_point ())
+	# IMPORTANT Determine which way the player is going. Make sure that pressing both left and right at the same time does nothing.
+	if (horizontal_lock_timer > 0):	# If movement is not possible...
+			pass# Make sure the movement variables are set to false.
+			# FIXME: Does jumping need to be in this block?
+	var floor_rays = [$FloorDetectCenter, $FloorDetectLeft, $FloorDetectRight]
+	for ray in floor_rays:
+#		# IMPORTANT FIXME: This is pretty hacky, so need to find a better way of handling all this.
+		is_player_on_floor = ray.is_colliding ()
+		if (is_player_on_floor && abs (velocity.x) >= 0.05):
+#			printerr (ray.name)
+			ground_normal = ray.get_collision_normal ()
+#			printerr (ray.name, " ", ground_normal)	# FOR DEBUGGING ONLY. Print which ray is colliding.
+			break
+	
+	var hitting_floor = (is_player_on_floor && !was_player_on_floor)
+	was_player_on_floor = is_player_on_floor
+	
+	var run_speed = (ground_speed if is_player_on_floor else velocity.x)
+	
+	
+	var ground_angle = (UP.angle_to (ground_normal))
+	
+	# FIXME: This piece of code seems to not be called at all!
+	if ((ground_angle >= (PI/2.01) || ground_angle <= (-PI/2.01)) && ground_speed < fall):
+		is_player_on_floor = false
+		ground_speed = 0  
+		horizontal_lock_timer = 0.5
+
+	if (is_player_on_floor):	# Major condition, determines whether we're going by groundspeed or traditional
+		# Vector that points "forward" along the ground that Sonic stands on.
+		var ground_speed_vector = ground_normal.rotated (PI/2)
+		if !(States & JUMP_STATE):
+			GroundSpeedometer(run_speed)
+		if (hitting_floor):
+			(States &= ~JUMP_STATE)
+			(States |= IDLE_STATE)
+			CheckStateToPlaySprite(States)
+			# If we've landed on floor, recalculate ground_speed.
+			ground_speed = velocity.dot (ground_speed_vector)
+			rotation = ground_angle
+#			printerr (ground_angle)
+#			move_and_collide ((-400) * ground_normal * delta)
+			move_and_slide ((-400) * ground_normal, UP)
```

Comment:
This line has always troubled me. It's been here since before I joined the project, and it doesn't really make sense. Things go awry if it's removed though, but clearly something is up enough to make it necessary.